p2p/src/simulated/network.rs

```diff
@@ -115,58 +257,70 @@ impl<E: Spawner + Rng + Clock> Network<E> {
+                    Err(err) => {
+                        let result = result.send(Err(err));
+                        if let Err(err) = result {
+                            error!(?err, "failed to send register err to oracle");
+                        }
+                        return;
                     }
-                    return;
-                }
+                };
+
+                // Create a sender that allows sending messages to the network for a certain channel
+                let sender = Sender::new(
+                    self.runtime.clone(),
+                    public_key,
+                    channel,
+                    self.max_size,
+                    self.sender.clone(),
+                );
 
-                // Initialize agent channel
-                let (sender, receiver) = mpsc::unbounded();
-                entry.insert(channel, (max_size, sender));
-                let result = result.send(Ok((
-                    Sender::new(
-                        self.runtime.clone(),
-                        public_key,
-                        channel,
-                        max_size,
-                        self.sender.clone(),
-                    ),
-                    Receiver { receiver },
-                )));
+                // Return values via callback
+                let result = result.send(Ok((sender, receiver)));
                 if let Err(err) = result {
                     error!(?err, "failed to send register ack to oracle");
                 }
             }
-            ingress::Message::Deregister { public_key, result } => {
-                if self.peers.remove(&public_key).is_none() {
-                    if let Err(err) = result.send(Err(Error::PeerMissing)) {
-                        error!(?err, "failed to send deregister err to oracle");
-                    }
-                    return;
-                }
-                if let Err(err) = result.send(Ok(())) {
-                    error!(?err, "failed to send deregister ack to oracle");
-                }
-            }
             ingress::Message::AddLink {
                 sender,
                 receiver,
                 sampler,
                 success_rate,
                 result,
             } => {
-                self.links
-                    .entry(sender)
-                    .or_default()
-                    .insert(receiver, (sampler, success_rate));
-                if let Err(err) = result.send(()) {
+                let peer = match self.peers.get(&sender) {
+                    Some(peer) => peer,
+                    None => {
+                        let result = result.send(Err(Error::PeerMissing));
+                        if let Err(err) = result {
+                            error!(?err, "failed to send add link err to oracle");
+                        }
+                        return;
+                    }
+                };
+
+                let link = Link::new(
```

Comment:
You should ensure there isn't already a link for this `sender/receiver` before creating this